lib/SourceMapDevToolPlugin.js

```diff
@@ -38,128 +38,126 @@ SourceMapDevToolPlugin.prototype.apply = function(compiler) {
 		compilation.plugin("after-optimize-chunk-assets", function(chunks) {
 			var allModules = [];
 			var allModuleFilenames = [];
-			var tasks = [];
-			chunks.forEach(function(chunk) {
-				chunk.files.filter(ModuleFilenameHelpers.matchObject.bind(undefined, options)).map(function(file) {
-					var asset = this.assets[file];
-					if(asset.__SourceMapDevToolData) {
-						var data = asset.__SourceMapDevToolData;
-						for(var cachedFile in data) {
-							this.assets[cachedFile] = data[cachedFile];
-							if(cachedFile !== file)
-								chunk.files.push(cachedFile);
+			function getSourceAndMap(asset) {
+				var source;
+				var sourceMap;
+				if(asset.sourceAndMap) {
+					var sourceAndMap = asset.sourceAndMap(options);
+					sourceMap = sourceAndMap.map;
+					source = sourceAndMap.source;
+				} else {
+					sourceMap = asset.map(options);
+					source = asset.source();
+				}
+
+				return {source: source, sourceMap: sourceMap};
+			}
+			function getModulesWithFilenames(sourceMap) {
+				var modules = sourceMap.sources.map(function(source) {
+					var module = compilation.findModule(source);
+					return module || source;
+				});
+				var moduleFilenames = modules.map(function(module) {
+					return ModuleFilenameHelpers.createFilename(module, moduleFilenameTemplate, requestShortener);
+				});
+
+				return {modules: modules, moduleFilenames: moduleFilenames}
+			}
+
+			var outputPath = this.getPath(compiler.outputPath);
+			compiler.outputFileSystem.mkdirp(outputPath, function() {
+				chunks.forEach(function(chunk) {
+					var files = chunk.files.filter(ModuleFilenameHelpers.matchObject.bind(undefined, options))
+					files.forEach(function(file) {
+						var {source, sourceMap} = getSourceAndMap(this.assets[file]);
```

Comment:
> `var {source, sourceMap}` 

I think this syntax isn't supported natively till node ~6. Not sure what node versions webpack tries to support.